src/main/java/com/google/devtools/build/lib/remote/merkletree/InputTree.java

```diff
@@ -0,0 +1,390 @@
+import com.google.devtools.build.lib.vfs.Symlinks;
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.SortedMap;
+import java.util.TreeMap;
+
+/**
+ * Intermediate tree representation of a list of lexicographically sorted list of files. Each
+ * node in the tree represents either a directory or file.
+ */
+class InputTree {
+
+  interface Visitor {
+    void visitDirectory(PathFragment dirname, List<FileNode> files, List<DirectoryNode> dirs);
+  }
+
+  public abstract static class Node {
+    private final String pathSegment;
+
+    public Node(String pathSegment) {
+      this.pathSegment = Preconditions.checkNotNull(pathSegment);
+    }
+
+    public String getPathSegment() {
+      return pathSegment;
+    }
+
+    @Override
+    public int hashCode() {
+      return pathSegment.hashCode();
+    }
+
+    @Override
+    public boolean equals(Object o) {
+      if (o == this) {
+        return true;
+      }
+      if (!(o instanceof Node)) {
+        return false;
+      }
+      Node other = (Node) o;
+      return pathSegment.equals(other.pathSegment);
+    }
+  }
+
+  public static class FileNode extends Node {
+    private final ActionInput input;
+    private final Digest digest;
+
+    public FileNode(String pathSegment, ActionInput input, Digest digest) {
+      super(pathSegment);
+      this.input = Preconditions.checkNotNull(input, "input");
+      this.digest = Preconditions.checkNotNull(digest, "digest");
+    }
+
+    public Digest getDigest() {
+      return digest;
+    }
+
+    public ActionInput getActionInput() {
+      return input;
+    }
+
+    @Override
+    public int hashCode() {
+      return Objects.hash(super.hashCode(), input, digest);
+    }
+
+    @Override
+    public boolean equals(Object o) {
+      if (o == this) {
+        return true;
+      }
+      if (!(o instanceof FileNode)) {
+        return false;
+      }
+      FileNode other = (FileNode) o;
+      return super.equals(o) && input.equals(other.input) && digest.equals(other.digest);
+    }
+  }
+
+  public static class DirectoryNode extends Node {
+    private final List<Node> children = new ArrayList<>();
+
+    public DirectoryNode(String pathSegment) {
+      super(pathSegment);
+    }
+
+    public void addChild(Node child) {
+      children.add(Preconditions.checkNotNull(child, "child"));
+    }
+
+    @Override
+    public int hashCode() {
+      return Objects.hash(super.hashCode(), children.hashCode());
+    }
+
+    @Override
+    public boolean equals(Object o) {
+      if (o == this) {
+        return true;
+      }
+      if (!(o instanceof DirectoryNode)) {
+        return false;
+      }
+      DirectoryNode other = (DirectoryNode) o;
+      return super.equals(other) && children.equals(other.children);
+    }
+  }
+
+  private final Map<PathFragment, DirectoryNode> tree;
```

Comment:
I wonder if there's a way to just take a plain list of ActionInputs and efficiently build the merkle tree directly from that. This is using a map, not a list, and the DirectoryNode instances seem to contain all nested elements.

Reply:
I believe there is for a lexicographically sorted list and it should be fairly easy to implement. However, what `InputTree` mostly is for is to expand directory inputs. I think after this change we could move the directory expansion to the `SpawnInputExpander` and after that we can build a merkle tree directly from a list of `ActionInputs`.